fairscale/optim/oss.py

```diff
@@ -310,12 +297,116 @@ def consolidate_state_dict(self, recipient_rank: int = 0) -> None:
+        """
+
+        if len(self._all_states) == 0:
+            raise RuntimeError(
+                "Optimizer state has not been consolidated on this rank. \
+                Please call `consolidate_state_dict()` on all ranks beforehand if you meant to save the global state"
+            )
+
+        # Unify the shard states and the state that pytorch would expect, given the model.
+        # Indexation needs several redirections, since each shard only knows a limited scope of the model
+        # - get the pytorch compliant parameter indexing
+        state_dict = super().state_dict()
+
+        # - get an id map which links the parameter id to the index in the reference state
+        global_id_map = {id(p): i for i, p in enumerate(chain(*(g["params"] for g in self.param_groups)))}
+
+        # - go through the per-shard states, which are all indexed locally
+        for rank, s in enumerate(self._all_states):
+            # -- match the local indexing and the global partition, update the corresponding saved state globally
+            for local_pg, global_pg in zip(s["param_groups"], self.partition_parameters()[rank]):
+                local_index_to_param_id = {
+                    i_param: id(global_pg["params"][i]) for i, i_param in enumerate(local_pg["params"])
+                }
+
+                for local_param_index in local_pg["params"]:
+                    # Update the state, if any
+                    if local_param_index in s["state"].keys():
+                        global_id = global_id_map[local_index_to_param_id[local_param_index]]
+                        state_dict["state"][global_id] = s["state"][local_param_index]
+
+        return state_dict
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        """Restore the global parameter groups as well as the shard.
+
+        Arguments:
+            state_dict (dict): optimizer state. Should be an object returned
+                from a call to :meth:`state_dict`
+        """
+
+        # NOTE: PyTorch 1.5 does not index linearly but with the id(params) at saving time
+        # we work around that here by using the fact that the params are ordered as in the param_groups
+
+        # Param index to param map
+        global_id_map = {i: p for i, p in enumerate(chain(*(g["params"] for g in self.param_groups)))}
+
+        # Prune the state_dict from the states which this rank does not own, then normal base load
+        for i_param, key in enumerate(state_dict["state"].keys()):
+            # Check that this rank owns this param, if not remove from the state
+            param = global_id_map[i_param]
+            if self.param_to_rank[param] != self.rank:
+                state_dict["state"][key] = None
+
+        super().load_state_dict(state_dict)
+
+        # Set the sharded optimizer state.
+        # Keep the original type (not respected by PyTorch which casts to the model type)
+        for k, (_, v) in enumerate(state_dict["state"].items()):
```

Comment:
I think this loop can be combined with the loop above and would result in simpler code.

Reply:
done in a new revision, I'm not sure whether it's simpler to understand but it does work and it's more concise indeed